### PR TITLE
Add EvolveNewtonianEuler Executable

### PR DESCRIPTION
**Pull request by carmaza**
## Proposed changes

The order is

- Add missing members to `NewtonianEuler` system.
- Add NewtonianEuler Initializer. 
  This commit adds two "new" files that should be replaced by something that doesn't duplicate code:
  + `src/Evolution/Conservative/UpdatePrimitivesNoRecovery.hpp`, which should be deleted in favor of making `UpdatePrimitives.hpp`handle systems without a list of prim from cons recovery methods.
  + `src/Evolution/Initialization/InterfaceNewtonianEuler.hpp`, which should be deleted in favor of making  `Interface.hpp` handle systems without GR variables, as well as making it capable of initializing system-specific compute items on the interface.

  In addition, `src/Evolution/Systems/NewtonianEuler/Initialize.hpp`  is also mostly a copy-paste of other systems, so we should somehow factor that out too (maybe in a future PR in order to deal with all systems at once).
- Add EvolveNewtonianEuler Executable.
 (2-3D for now because there is no 1D initial data in develop yet!!!)

Any suggestions to deal with the code duplication in the initialization would be appreciated!

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Review by kidder — COMMENTED** [review]
this test gives output which needs to either turned off or removed (see Poisson executable)

Otherwise, the last 3 commits look good (group review with @nilsdeppe @wthrowe )

**Comment by carmaza** [comment]
I just removed the options of `EventsAndTriggers` in the input files, as done in other Evolution systems.

**Review by kidder — DISMISSED** [review]

File: src/Evolution/Systems/NewtonianEuler/System.hpp [hunk 1]
```diff
@@ -5,15 +5,62 @@
...
+#include "Evolution/Conservative/ConservativeDuDt.hpp"
+#include "Evolution/Systems/NewtonianEuler/Characteristics.hpp"
+#include "Evolution/Systems/NewtonianEuler/ConservativeFromPrimitive.hpp"
+#include "Evolution/Systems/NewtonianEuler/Fluxes.hpp"
+#include "Evolution/Systems/NewtonianEuler/PrimitiveFromConservative.hpp"
+#include "Evolution/Systems/NewtonianEuler/Tags.hpp"
+#include "PointwiseFunctions/Hydro/Tags.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace Tags {
```
Comment (nilsdeppe) [inline src/Evolution/Systems/NewtonianEuler/System.hpp:20]: `cond` and `endcond` around this please.

File: src/Evolution/Systems/NewtonianEuler/System.hpp [hunk 2]
```diff
@@ -5,15 +5,62 @@
...
-template <size_t Dim>
+template <size_t Dim, typename EquationOfStateType>
 struct System {
   static constexpr bool is_in_flux_conservative_form = true;
   static constexpr bool has_primitive_and_conservative_vars = true;
   static constexpr size_t volume_dim = Dim;
+  static constexpr size_t thermodynamic_dim =
+      EquationOfStateType::thermodynamic_dim;
+
+  // Compute item for pressure not currently implemented in SpECTRE,
```
Comment (nilsdeppe) [inline src/Evolution/Systems/NewtonianEuler/System.hpp:38]: Why isn't a compute item added? Why is this different from GRMHD?

File: src/Evolution/Systems/NewtonianEuler/Initialize.hpp [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+template <class T, class = cpp17::void_t<>>
+struct has_analytic_solution_alias : std::false_type {};
+template <class T>
+struct has_analytic_solution_alias<T,
+                                   cpp17::void_t<typename T::analytic_solution>>
+    : std::true_type {};
+}  // namespace detail
+
+template <size_t Dim>
+struct Initialize {
```
Comment (nilsdeppe) [inline src/Evolution/Systems/NewtonianEuler/Initialize.hpp]: Can you please file an issue to remove this once PDALs are in? This code is almost identical to the GRMHD case and I think we should figure out how to not do that once we have PDALs :)

File: src/Evolution/Initialization/InterfaceNewtonianEuler.hpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
+// Distributed under the MIT License.
```
Comment (nilsdeppe) [inline src/Evolution/Initialization/InterfaceNewtonianEuler.hpp]: Two points on this:

- Why is this necessary? It should be clearly stated in the documentation why the current generic one isn't used.
- This should be in the NE directory since it's system specific.

File: src/Evolution/Initialization/InterfaceNewtonianEuler.hpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+///                                Tags::UnnormalizedFaceNormal<Dim>>`
+///   * Tags::InterfaceComputeItem<Directions,
+///                                typename System::template magnitude_tag<
+///                                    Tags::UnnormalizedFaceNormal<Dim>>>
+///   * `Tags::InterfaceComputeItem<
+///         Directions, Tags::Normalized<Tags::UnnormalizedFaceNormal<Dim>>>`
+/// - Removes: nothing
+/// - Modifies: nothing
+template <typename System>
+struct Interface {
```
Comment (nilsdeppe) [inline src/Evolution/Initialization/InterfaceNewtonianEuler.hpp]: Just make the generic one allow you to add compute face and ext tags, and while it's being changed it also should only add prims if the system has prims.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/NewtonianEuler/System.hpp (hunk 2 above)
Comment (carmaza) [reply src/Evolution/Systems/NewtonianEuler/System.hpp:38]: If I'm not wrong, one would like to add the pressure as a generic compute item, where the function to compute it is retrieved from the equation of state. What is done in this system *is* essentially the same that was done in GRMHD. 

If we implement the compute item, we might also want to do it for all hydro systems at once. Definitely something for a different PR.

File: src/Evolution/Systems/NewtonianEuler/Initialize.hpp (hunk 3 above)
Comment (carmaza) [reply src/Evolution/Systems/NewtonianEuler/Initialize.hpp]: Sure, I'll also mention that it should be removed from GRMHD (and possibly from GH, if it gets merged soon).

File: src/Evolution/Initialization/InterfaceNewtonianEuler.hpp (hunk 5 above)
Comment (carmaza) [reply src/Evolution/Initialization/InterfaceNewtonianEuler.hpp]: Yes, that's what I wanted too, but I'm afraid I'll need a more detailed explanation on how to do this :) Otherwise, I'll move this file to the NE directory.

File: src/Evolution/Systems/NewtonianEuler/System.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/Evolution/Systems/NewtonianEuler/System.hpp:38]: I think the reason we can't currently have a compute item is we need PDALs to get const global cache things into the DataBox. Future problem it is :)

File: src/Evolution/Initialization/InterfaceNewtonianEuler.hpp (hunk 5 above)
Comment (nilsdeppe) [reply src/Evolution/Initialization/InterfaceNewtonianEuler.hpp]: Okay, maybe we can talk about this in person tomorrow (Wednesday)?

File: src/Evolution/Systems/NewtonianEuler/Initialize.hpp (hunk 3 above)
Comment (nilsdeppe) [reply src/Evolution/Systems/NewtonianEuler/Initialize.hpp]: Thanks!

File: src/Evolution/Initialization/InterfaceNewtonianEuler.hpp (hunk 5 above)
Comment (nilsdeppe) [reply src/Evolution/Initialization/InterfaceNewtonianEuler.hpp]: Here are the metafunctions and code that allows you to pass in a typelist of tags where the first tag will be substituted in with the `Direction` tag.
```cpp
template <typename DirectionsTag, typename VarsTag>
struct Slice {};

template <typename T>
struct SoundSpeed {};

template <template <class...> class F, typename... Ts>
struct interface_holder {
  template <typename Dir>
  using f = F<Dir, Ts...>;
};

template <typename InterfaceHolder, typename Direction>
struct add_direction_tag {
  using type = typename InterfaceHolder::template f<Direction>;
};

using tags = tmpl::list<interface_holder<Slice, SoundSpeed<double>>>;

struct Dir {};

using result = tmpl::transform<tags, add_direction_tag<tmpl::_1, Dir>>;

td<result> aoeu;
```
The compiler output is:
```
test.cpp:167:12: error: implicit instantiation of undefined template
      'td<brigand::list<Slice<Dir, SoundSpeed<double> > > >'
td<result> aoeu;
```
which shows that the transformation was done correctly. The `interface_holder` class can probably just go into the `Initialization` namespace, while the metafunction `add_direction_tag` can go inside the `Initialization::Interface` class.

**Comment by carmaza** [comment]
Rebased on develop. Updated commit list is:

- Add missing members to system
- Add initialization:
  + I no longer need to add an "UpdatePrimitivesNoRecovery" file, as `MutateApply` can handle that in the executable (although I needed to add an extra fixup commit to make it work; see below). 
  + Initialization for the system is still a copy-paste of other systems. I can file an issue to remove that (as originally requested), or let it be for now and deal with that after pdals are in (not sure how but Im open to learn).
  + Initialization of the interface is specific for this system, and it's a copy-paste from the generic Interface.hpp file, because I need to add compute items specific to this system. (I did manage to make the generic file handle primitives and spacetime variables as optional arguments, though. Maybe that could be added in a different PR?)
- Fixup commit to add a template paratemer to `MutateApply` (Was this class actually tested? I can't find anywhere in SpECTRE where it's being used.)

- Add EvolveNewtonianEuler executable with 2 and 3D input files. 1D coming soon along with a 1D analytic solution (Riemann problem is the best candidate.)

**Comment by carmaza** [comment]
Rebased on develop. The commit order is :

- add missing members to system
- fix ctest labels in the recently moved ParallelAlgorithms tests (let me know if this isn't needed)
- add support for initializing ComputeTags
- add NewtonianEuler executable updated to use PDALs.

File: src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+                                     initialization_actions>,
+
+              Parallel::PhaseActions<
+                  Phase, Phase::InitializeTimeStepperHistory,
+                  tmpl::flatten<tmpl::list<SelfStart::self_start_procedure<
+                      compute_rhs, update_variables>>>>,
+
+              Parallel::PhaseActions<
+                  Phase, Phase::RegisterWithObserver,
+                  tmpl::list<Actions::AdvanceTime,
```
Comment (nilsdeppe) [inline src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp]: With #1604 merged this `AdvanceTime` can be removed

File: src/ParallelAlgorithms/Initialization/AddComputeTags.hpp [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+/// \cond
+namespace Parallel {
+template <typename Metavariables>
+class ConstGlobalCache;
+}  // namespace Parallel
+/// \endcond
+
+namespace Initialization {
+namespace Actions {
+template <typename ComputeTagsList>
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Initialization/AddComputeTags.hpp]: Could you please write dox for this following the format that other actions use? It looks like the other action in this directory is in the `ActionsGroup`.

File: tests/Unit/ParallelAlgorithms/Initialization/Test_AddComputeTags.cpp [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+#include "tests/Unit/TestingFramework.hpp"
+
+#include <cstddef>
+#include <limits>
+#include <random>
+#include <string>
+#include <tuple>
+#include <utility>
+
+#include <iostream>
```
Comment (nilsdeppe) [inline tests/Unit/ParallelAlgorithms/Initialization/Test_AddComputeTags.cpp]: Whoops, I think this stuck around from debugging and can be removed now :)

**Review by nilsdeppe — COMMENTED** [review]
Just a couple minor comments. I'm happy for you to squash things immediately.

This is really nice. Good job! :)

File: src/ParallelAlgorithms/Initialization/AddComputeTags.hpp (hunk 7 above)
Comment (carmaza) [reply src/ParallelAlgorithms/Initialization/AddComputeTags.hpp]: so should this header go in `ParallelAlgorithms/Initialization/Actions` then?

Comment (nilsdeppe) [reply src/ParallelAlgorithms/Initialization/AddComputeTags.hpp]: hmm yea that seems most consistent

Comment (carmaza) [reply src/ParallelAlgorithms/Initialization/AddComputeTags.hpp]: Changed the location and added some dox. Let me know if something's wrong or if there are better words for it.

**Review by nilsdeppe — DISMISSED** [review]

File: src/Evolution/Executables/NewtonianEuler/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  DiscontinuousGalerkin
+  DomainCreators
+  Hydro
+  IO
+  Informer
+  LinearOperators
+  MathFunctions
+  NewtonianEuler
+  NewtonianEulerSolutions
+  Limiters
```
Comment (fmahebert) [inline src/Evolution/Executables/NewtonianEuler/CMakeLists.txt]: Alphabetize

File: src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+              Initialization::option_tags<initialization_actions>>>>;
+
+  using const_global_cache_tag_list =
+      tmpl::list<analytic_solution_tag,
+                 OptionTags::TypedTimeStepper<tmpl::conditional_t<
+                     local_time_stepping, LtsTimeStepper, TimeStepper>>,
+                 OptionTags::EventsAndTriggers<events, triggers>>;
+
+  static constexpr OptionString help{
+      "Evolve the Newtonian Euler system in consevative form.\n\n"};
```
Comment (fmahebert) [inline src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp]: Typo in "consevative"

**Review by fmahebert — COMMENTED** [review]
Two minor observations, I don't mind if you prefer to fix these in a subsequent PR 🙂

**Comment by carmaza** [comment]
thanks @fmahebert they should be fixed now.

**Review by fmahebert — APPROVED** [review]
Speaking just to the Newtonian Euler part (not the ComputeTag init part): LGTM

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]
